tritium-caffeine/src/main/java/com/palantir/tritium/metrics/caffeine/CacheStats.java

```diff
@@ -61,17 +67,41 @@ public static CacheStats of(TaggedMetricRegistry taggedMetricRegistry, @Safe Str
         return new CacheStats(CacheMetrics.of(taggedMetricRegistry), name);
     }
 
+    /**
+     * Registers additional metrics for a Caffeine cache.
+     *
+     * @param cache cache for which to register metrics
+     * @return the given cache instance
+     */
+    public <K, V, C extends Cache<K, V>> C register(C cache) {
+        metrics.estimatedSize().cache(name).build(cache::estimatedSize);
+        metrics.weightedSize().cache(name).build(() -> cache.policy()
+                .eviction()
+                .flatMap(e -> e.weightedSize().stream().boxed().findFirst())
+                .orElse(null));
+        metrics.maximumSize().cache(name).build(() -> cache.policy()
+                .eviction()
+                .map(Policy.Eviction::getMaximum)
+                .orElse(null));
```

Comment:
Unless the `CaffeineCacheStats` implementation which returns 0 and -1 respectively for these guages when they are not present, this implementation just returns `null` so the metric is omitted. This feels objectively more correct and avoids confusion when the weighted size appears to be 0.

Reply:
@carterkozak do you recall why we used -1? I think there was some reason I didn't return null previously (should have left a comment). I need to take a look at the metric logging output side, but I think null should be fine.

Reply:
I think we might have some logging for non-numeric values, however if that flags null, we should update the logging to silently ignore nulls (as opposed to strings/booleans/etc)

Reply:
-1 was a hack, null is more correct

https://github.com/palantir/tritium/pull/268#pullrequestreview-239201095